### ISSUE TITLE
Using the contour plot to visualize the prediction of classes

### ISSUE THREAD

**Issue by shuaigezhu** [shuaigezhu/star_galaxy_classification#24]

**Comment by shuaigezhu** [comment]
![Screen Shot 2020-03-25 at 11 33 14 AM](https://user-images.githubusercontent.com/47727949/77572600-81c6fd00-6e8c-11ea-89a7-67651903d2d3.png)

Currently, I plot all the pixels which are over 80% of the prediction of three classes. Green is the quasar, red is the star and blue is the galaxy. Today I am going to use **contour plot** to show the prediction of the three classes.

**Comment by kmyi** [comment]
Can you show also an overlay comparing between ground-truth vs what the network is doing?
--
Kwang Moo YI
Assistant Professor
Dept. of Computer Science
University of Victoria
+1 (250) 472 5837

**Comment by shuaigezhu** [comment]
![Screen Shot 2020-03-25 at 2 14 55 PM](https://user-images.githubusercontent.com/47727949/77586271-20f6ef00-6ea3-11ea-95b1-3c6805975d45.png)
![Screen Shot 2020-03-25 at 2 14 40 PM](https://user-images.githubusercontent.com/47727949/77586275-23594900-6ea3-11ea-925a-3764f9323df6.png)
![Screen Shot 2020-03-25 at 2 14 24 PM](https://user-images.githubusercontent.com/47727949/77586278-23f1df80-6ea3-11ea-9b43-d0c39b8c37c4.png)
![Screen Shot 2020-03-25 at 2 14 11 PM](https://user-images.githubusercontent.com/47727949/77586279-248a7600-6ea3-11ea-8ed1-bba998fecf81.png)


I make a larger circle to indicate the class of the object. Similarly, green is the quasar, red is the star and blue is the galaxy. Since we have no information(ground truth) for lots of objects, you can see some objects don't have a larger circle. Currently, the model easily predicts star or quasar as a galaxy and you can notice this from above screenshots. The reason is there are so many galaxies and fewer stars and quasars in the catalog and training set. Thus, I think I need to balance input data somehow.

**Comment by kmyi** [comment]
Can you simply display only the ground-truth on one side and the label on the other for each channel? E.g. you don’t have to display multiple channels, you can simply display GT vs what you have. Also, display for each object type independently. What we want to see here is to determine how good we are. Try to visualize that.

--
Kwang Moo YI
Assistant Professor
Dept. of Computer Science
University of Victoria
+1 (250) 472 5837

**Comment by shuaigezhu** [comment]
I will display each object type independently by contour plot. I count the precision for each class, and here is the print.
print:
there are 515 objects in the test set
there are 113 quasars totally, and 92 of them are predicted as the quasar
there are 84 stars totally, and 38 of them are predicted as the star
there are 297 galaxies totally, and 278 of them are predicted as the galaxy
overall precision is 81.74757281553397
end

These are only for the objects which have the label. I found out the labeled pixel in the label and found out the corresponding pixel in the prediction, then I compare the class with the highest probability.

**Comment by shuaigezhu** [comment]
> I will display each object type independently by contour plot. I count the precision for each class, and here is the print.
> print:
> there are 515 objects in the test set
> there are 113 quasars totally, and 92 of them are predicted as the quasar
> there are 84 stars totally, and 38 of them are predicted as the star
> there are 297 galaxies totally, and 278 of them are predicted as the galaxy
> overall precision is 81.74757281553397
> end
> 
> These are only for the objects which have the label. I found out the labeled pixel in the label and found out the corresponding pixel in the prediction, then I compare the class with the highest probability.

**Comment by shuaigezhu** [comment]
> > I will display each object type independently by contour plot. I count the precision for each class, and here is the print.
> > print:
> > there are 515 objects in the test set
> > there are 113 quasars totally, and 92 of them are predicted as the quasar
> > there are 84 stars totally, and 38 of them are predicted as the star
> > there are 297 galaxies totally, and 278 of them are predicted as the galaxy
> > overall precision is 81.74757281553397
> > end
> > These are only for the objects which have the label. I found out the labeled pixel in the label and found out the corresponding pixel in the prediction, then I compare the class with the highest probability.

**Comment by shuaigezhu** [comment]
I will display each object type independently by contour plot. Currently, I count the precision for each class, and here is the print.
print:
there are 515 objects in the test set
there are 113 quasars totally, and 92 of them are predicted as the quasar
there are 84 stars totally, and 38 of them are predicted as the star
there are 297 galaxies totally, and 278 of them are predicted as the galaxy
overall precision is 81.74757281553397
end

These are only for the objects which have the label. I found out the labeled pixel in the label and found out the corresponding pixel in the prediction, then I compare the class with the highest probability.

**Comment by kmyi** [comment]
I don't understand how you could compute precision, whereas your problem is really a segmentation problem. Shouldn't you compute the dice score?

**Comment by kmyi** [comment]
Also, if you want to use precision, you need to compute what people call mean average precision, since you have multiple classes.

**Comment by shuaigezhu** [comment]
ok, let me take an example to explain precision here.
label pixel in y is [0.8, 0.2, 0, 0], index 0 has the highest probability so I take it as my ground truth.
the corresponding pixel in logits is [0.7, 0.2, 0.1], index 0 has the highest probability so I take it as the predicted class. The ground truth class and prediction class are the same, then I set the prediction for this object is true.

**Comment by shuaigezhu** [comment]
yes, this is a segmentation problem, but how do I label the class of objects in prediction? As the image above, you can see, for one object, some parts are predicted as stars and some parts are predicted as galaxies or quasars. Since we use the central pixel to train the model, maybe the central area of the objects in prediction can make more sense?

**Comment by shuaigezhu** [comment]
The central pixel means the label pixel in the label image. Each object in training has exactly one label pixel and the position of the pixel in on the center of object roughly.

**Comment by kmyi** [comment]
One thing you could do is simply apply a gaussian blob on these object locations to create a “fake groudn truth” which would then give you some idea on what the performance would be. You could also visualize or compute the metric as you did with a circle on top, which I guess would be more clear if you simply do this on a class basis.
--
Kwang Moo YI
Assistant Professor
Dept. of Computer Science
University of Victoria
+1 (250) 472 5837

**Comment by kmyi** [comment]
> ok, let me take an example to explain precision here.
> label pixel in y is [0.8, 0.2, 0, 0], index 0 has the highest probability so I take it as my ground truth.
> the corresponding pixel in logits is [0.7, 0.2, 0.1], index 0 has the highest probability so I take it as the predicted class. The ground truth class and prediction class are the same, then I set the prediction for this object is true.

The question I have for this is "then how do you incorporate non-object pixels" in your evaluation?

**Comment by shuaigezhu** [comment]
> > ok, let me take an example to explain precision here.
> > label pixel in y is [0.8, 0.2, 0, 0], index 0 has the highest probability so I take it as my ground truth.
> > the corresponding pixel in logits is [0.7, 0.2, 0.1], index 0 has the highest probability so I take it as the predicted class. The ground truth class and prediction class are the same, then I set the prediction for this object is true.
> 
> The question I have for this is "then how do you incorporate non-object pixels" in your evaluation?

Actually, I just apply a filter to select the pixels above a certain value. Most non-object pixels probabilities are close to 0.25, so for example, I use the filter to only select and evaluate the label pixels above 0.5.

**Comment by shuaigezhu** [comment]
![Screen Shot 2020-04-01 at 10 21 53 AM](https://user-images.githubusercontent.com/47727949/78167493-6c5d5000-7403-11ea-880e-1bb453b882d5.png)

It looks really weird to use Matplotlib to plot contour.
